File: LinuxTuner/MountManager/PopupWindow.py
from PyQt4 import Qt
import Const

class PopupWindow(Qt.QDialog):
	### Window wich show after press some buttons to show info about e.g. some option ##
	def __init__(self,doc_center,parent = None):
		Qt.QDialog.__init__(self,parent)
				
		self.setWindowFlags(Qt.Qt.Popup)
		self.resize(300,200)
		
		self.doc_center = doc_center
		
		self.cursor = Qt.QCursor()
		
		self.close_button = Qt.QToolButton()
		self.close_button.setIcon(Qt.QIcon(Const.Icons_path + "close_popup.png"))
		self.connect(self.close_button,Qt.SIGNAL("clicked()"),self.hide)
		
		self.title_label = Qt.QLabel()
		
		self.text_browser = Qt.QTextBrowser()
		
		self.top_layout = Qt.QHBoxLayout()
		self.top_layout.addWidget(self.title_label)
		self.top_layout.addStretch()
		self.top_layout.addWidget(self.close_button)
		
		self.main_layout = Qt.QVBoxLayout()
		self.main_layout.setMargin(2)
		self.main_layout.setSpacing(2)
		self.main_layout.addLayout(self.top_layout)
		self.main_layout.addWidget(self.text_browser)
		
		self.setLayout(self.main_layout)
	
	def setData(self,popup_title,popup_text,doc_mark = None,width = 300,height = 200):
		### Set text in text browser ###
		self.resize(width,height)
		
		# Search documentation 
		if doc_mark != None:
			if self.doc_center.getInfoAboutFs(doc_mark) != None:
				popup_text += "<br>" + self.doc_center.getInfoAboutFs(doc_mark)
			elif self.doc_center.getGeneralInfo(doc_mark) != None:
				popup_text += "<br>" + self.doc_center.getGeneralInfo(doc_mark)
			elif self.doc_center.getCommonOptionInfo(doc_mark) != None:
				popup_text += "<br>" + self.doc_center.getCommonOptionInfo(doc_mark)
			elif self.doc_center.getDependsOnFsOptionInfo(doc_mark) != None:
				popup_text += "<br>" + self.doc_center.getDependsOnFsOptionInfo(doc_mark)
						
		
		
		# move window under cursor
		self.move(self.cursor.pos())
		
		self.title_label.setText("<b>" + popup_title + "</b>")
		
		if Qt.QString(popup_text).trimmed().isEmpty():
			popup_text = self.tr("There isn't documentation. Update package \"mountmanager-doc\" or report to developers at the site") + " " + Const.Site 
		
		self.text_browser.setHtml(popup_text)
		
		self.show()
		
		

File: trunk/lightlang-editor/src/PopupWindow.cpp
#include <QtGui/QToolButton>
#include <QtGui/QApplication>
#include <QtGui/QDesktopWidget>
#include <QtGui/QTextBrowser>
#include <QtGui/QLabel>
#include <QtGui/QVBoxLayout>
#include <QtGui/QHBoxLayout>
#include <QtGui/QCursor>
#include "PopupWindow.h"

PopupWindow::PopupWindow(QWidget *parent) : QDialog(parent) {

	setWindowFlags(Qt::Popup);

	wasChanges = false;
	header = new QLabel;

	browser = new QTextBrowser;
	connect(browser,SIGNAL(textChanged()),this,SLOT(textChanged()));

	cursor = new QCursor();
	
	closeButton = new QToolButton;
	closeButton->setAutoRaise(true);
	closeButton->setIcon(QIcon(":/icons/close.png"));
	connect(closeButton,SIGNAL(clicked()),this,SLOT(hide()));

	QHBoxLayout *topLayout = new QHBoxLayout;
	topLayout->addWidget(header);
	topLayout->addStretch();
	topLayout->addWidget(closeButton);
	
	QVBoxLayout *mainLayout = new QVBoxLayout;
	mainLayout->addLayout(topLayout);
	mainLayout->addWidget(browser);
	mainLayout->setMargin(3);
	mainLayout->setSpacing(3);

	setLayout(mainLayout);
	resize(300,200);
}

PopupWindow::~PopupWindow() {
	delete browser;
	delete closeButton;
	delete header;
	delete cursor;
}

void PopupWindow::setText(const QString &text) {
	browser->setHtml(text);
}

void PopupWindow::setHeaderText(const QString &headerText) {
	header->setText("<b>" + headerText + "</b>");
}

void PopupWindow::showPopup() {
	int x = 0;
	int y = 0;
	x = cursor->pos().x();
	y = cursor->pos().y();
	if (x < 0) x = 0;
	if (y < 0) y = 0;
	if (x + width() > QApplication::desktop()->width())
		x = QApplication::desktop()->width() - width() - 10;
	if (y + height() > QApplication::desktop()->height())
		y = QApplication::desktop()->height() - height() - 10;
	move(QPoint(x,y));
	show();
	browser->setFocus();
}

void PopupWindow::hideEvent(QHideEvent *) {
	if (wasChanges)
		emit (closedAfterChanges());
	wasChanges = false;
	emit (closed());
}

void PopupWindow::showEvent(QShowEvent *) {
	wasChanges = false;
}

void PopupWindow::setReadOnly(bool readOnly) {
	browser->setReadOnly(readOnly);
}

QString PopupWindow::getText() const {
	return browser->toPlainText();
}

void PopupWindow::textChanged() {
	wasChanges = true;
}
